design-documents/Remove-__sleep-and-__wakeup.md

```diff
@@ -0,0 +1,35 @@
+### Remove __sleep and __wakeup methods from our code
+#### Why?
+* Serializing objects requires classes to always consider serialization process
+and it is easy for multiple developers working on the same class to forget about it.
+* PHP serialization opens up our code to potential security issues (Remote code execution).
+* During serialization objects of classes that do not expect serialization may be serialized
+anyway causing unexpected behaviour.
+
+#### Current situation
+The only place we serialize objects (as far as I am aware of) is storing User and Acl objects
+in admin session for convenience during admin auth process. Indirect serialization caused
+us to add __sleep and __wakeup methods to ~33 classes.
+
+#### What has to be done
+##### Admin session
+I propose creating 2 interfaces in the Magento\Backend\Spi namespace:
```

Comment:
What would be a benefit in having these interfaces part of module API?

Reply:
There may be additional logic needed for 3rd party devs who extend User or Role classes, so they would implement their own hydrators. We can't just use getData() and setData() on those objects and expect that to be enough for everyone.

Reply:
We don't have anywhere Spi as part of namespace, I propose to discuss adding this convention separately. To add more data to User and Acl, you would need to modify User and Acl and hydrators. Can we make hydrators generic? Maybe to extract data, we can have toArray method and creation of the entity will be done using constructor (haven't look at the code, don't know if it's feasible)?